extsort.c

```diff
@@ -0,0 +1,250 @@
+                error("Full path to the temporary folder is too long\n");
+            strcat(tmp_path, "/bcftools-sort.XXXXXX");
+            es->tmp_prefix = strdup(tmp_path);
+        #else
+            es->tmp_prefix = strdup("/tmp/bcftools-sort.XXXXXX");
+        #endif
+    }
+}
+
+void extsort_set(extsort_t *es, extsort_opt_t key, void *value)
+{
+    if ( key==DAT_SIZE ) { es->dat_size = *((size_t*)value); return; }
+    if ( key==MAX_MEM )
+    {
+        es->max_mem = parse_mem_string(*((const char**)value));
+        if ( es->max_mem <=0 ) error("Could not parse the memory string, expected positive number: %s\n",*((const char**)value));
+        return;
+    }
+    if ( key==TMP_PREFIX ) { _init_tmp_prefix(es, *((const char**)value)); return; }
+    if ( key==FUNC_CMP ) { es->cmp = *((extsort_cmp_f*)value); return; }
+}
+
+extsort_t *extsort_alloc(void)
+{
+    extsort_t *es = (extsort_t*) calloc(1,sizeof(*es));
+    es->max_mem = 100e6;
+    return es;
+}
+void extsort_init(extsort_t *es)
+{
+    assert( es->cmp );
+    assert( es->dat_size );
+    if ( !es->tmp_prefix ) _init_tmp_prefix(es, NULL);
+    es->tmp_dat = malloc(es->dat_size);
+}
+
+void extsort_destroy(extsort_t *es)
+{
+    int i;
+    for (i=0; i<es->nblk; i++)
+    {
+        blk_t *blk = es->blk[i];
+        if ( blk->fd!=-1 ) close(blk->fd);
+        free(blk->fname);
+        free(blk->dat);
+        free(blk);
+    }
+    free(es->tmp_dat);
+    free(es->tmp_prefix);
+    free(es->blk);
+    khp_destroy(blk, es->bhp);
+    free(es);
+}
+
+static void _buf_flush(extsort_t *es)
+{
+    if ( !es->nbuf ) return;
+
+    qsort(es->buf, es->nbuf, sizeof(void*), es->cmp);
+
+    es->nblk++;
+    es->blk = (blk_t**) realloc(es->blk, sizeof(blk_t*)*es->nblk);
+    es->blk[es->nblk-1] = (blk_t*) calloc(1,sizeof(blk_t));
+    blk_t *blk = es->blk[es->nblk-1];
+    blk->es    = es;
+    blk->dat   = malloc(es->dat_size);
+    blk->fname = strdup(es->tmp_prefix);
+    #ifdef _WIN32
+        int i;
+        for (i=0; i<100000; i++)
+        {
+            memcpy(blk->fname,es->tmp_prefix,strlen(es->tmp_prefix));
+            mktemp(blk->fname);
+            blk->fd = _open(blk->fname, O_RDWR|O_CREAT|O_EXCL|O_BINARY|O_TEMPORARY, 0600);
+            if ( blk->fd==-1 )
+            {
+                if ( errno==EEXIST ) continue; 
+                error("Error: failed to open a temporary file %s\n",blk->fname);
+            }
+        }
+        if ( !blk->fd ) error("Error: failed to create a unique temporary file name from %s\n",es->tmp_prefix);
+        if ( fchmod(blk->fd,S_IRUSR|S_IWUSR)!=0 ) error("Error: failed to set permissions of the temporary file %s\n",blk->fname);
+    #else
+        if ( (blk->fd = mkstemp(blk->fname))==-1 )
+            error("Error: failed to open a temporary file %s\n",blk->fname);
+        if ( fchmod(blk->fd,S_IRUSR|S_IWUSR)!=0 ) error("Error: failed to set permissions of the temporary file %s\n",blk->fname);
```

Comment:
The call to `fchmod` should be replaced with a call to `_chmod` on the filename, instead of the descriptor.
`_chmod(blk->fname, S_IRUSR|S_IWUSR) `